File: src/alerts/gating.py
"""Глобальное решение «есть ли что слать»: events → regime → edge → cooldown.

Per-user silence НЕ проверяется здесь — это зона рассылки (analyze.py),
которая для каждого user'а смотрит его silence отдельно.

Pure-функция: на вход состояние, calendar, score, config; на выход —
GateDecision (allow/suppress + reason).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from src.analysis.scoring import Regime, regime_rank
from src.events.filter import Event, is_in_blackout
from src.state.store import State


@dataclass(frozen=True)
class GateDecision:
    allow: bool
    reason: str  # человекочитаемое объяснение для логов и /status


def decide(
    *,
    state: State,
    new_regime: Regime,
    edge_pct: float,
    events: list[Event],
    min_edge_pct: float,
    cooldown_hours: int,
    now: datetime | None = None,
) -> GateDecision:
    """Глобальное решение — слать ли сейчас алерт хоть кому-то.

    Порядок проверок (first-match wins для лога reason):
    1. event blackout
    2. regime ∈ {cooldown, watch}
    3. edge < min_edge_pct (если min_edge_pct > 0)
    4. regime равен/ниже последнего за cooldown_hours
    5. → разрешить
    """
    now = now or datetime.now(timezone.utc)

    in_blackout, event = is_in_blackout(now, events)
    if in_blackout and event is not None:
        return GateDecision(False, f"event blackout: {event.type} {event.title} at {event.ts.isoformat()}")

    if new_regime in ("cooldown", "watch"):
        return GateDecision(False, f"regime '{new_regime}' below alert threshold")

    if min_edge_pct > 0 and edge_pct < min_edge_pct:
        return GateDecision(False, f"edge {edge_pct:.2f}% < min {min_edge_pct:.2f}%")

    last = state.last_alert
    if last is not None:
        delta = now - _ensure_utc(last.ts)
        if delta < timedelta(hours=cooldown_hours):
            new_rank = regime_rank(new_regime)
            old_rank = regime_rank(_safe_regime(last.regime))
            if new_rank <= old_rank:
                return GateDecision(
                    False,
                    f"cooldown: same/lower regime ('{last.regime}' → '{new_regime}') "
                    f"within {cooldown_hours}h",
                )
            # else: апгрейд regime — пропускаем

    return GateDecision(True, f"regime '{new_regime}', edge {edge_pct:+.2f}%")


def compute_edge_pct(current_rate: float, baseline_median_30d: float | None) -> float:
    """edge = (current - baseline) / baseline * 100. Если baseline нет — 0."""
    if baseline_median_30d is None or baseline_median_30d == 0:
        return 0.0
    return (current_rate - baseline_median_30d) / baseline_median_30d * 100.0


def _ensure_utc(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def _safe_regime(value: str | None) -> Regime | None:
    if value in ("cooldown", "watch", "partial", "strong"):
        return value  # type: ignore[return-value]
    return None

File: worker/src/analyze/gating.ts
/**
 * Глобальное решение «слать ли алерт» — events → regime → edge → cooldown.
 * Port of src/alerts/gating.py.
 */
import type { AlertRecord } from "../state/schema";
import { type Candle, type Regime, regimeRank } from "./scoring";

export interface GateDecision {
  allow: boolean;
  reason: string;
}

export interface EventLike {
  type: string;
  title?: string;
  ts: string;
}

export interface GatingContext {
  newRegime: Regime;
  edgePct: number;
  /** Активное blackout-событие, или null. */
  blackoutEvent: EventLike | null;
  /** Последний отправленный алерт (для cooldown). */
  lastAlert: AlertRecord | null;
  /** Текущее время (ISO 8601 UTC). */
  now: string;
  minEdgePct: number;
  cooldownHours: number;
}

/**
 * Решает: слать ли алерт. Порядок first-match:
 *   1. event blackout
 *   2. regime ∈ {cooldown, watch}
 *   3. edge < min_edge_pct
 *   4. cooldown: same/lower regime within cooldown_hours
 *   5. → allow
 */
export function decide(ctx: GatingContext): GateDecision {
  if (ctx.blackoutEvent !== null) {
    const event = ctx.blackoutEvent;
    return {
      allow: false,
      reason: `event blackout: ${event.type} ${event.title ?? ""} at ${event.ts}`.trim(),
    };
  }
  if (ctx.newRegime === "cooldown" || ctx.newRegime === "watch") {
    return { allow: false, reason: `regime '${ctx.newRegime}' below alert threshold` };
  }
  if (ctx.minEdgePct > 0 && ctx.edgePct < ctx.minEdgePct) {
    return {
      allow: false,
      reason: `edge ${ctx.edgePct.toFixed(2)}% < min ${ctx.minEdgePct.toFixed(2)}%`,
    };
  }
  if (ctx.lastAlert !== null) {
    const deltaMs = new Date(ctx.now).getTime() - new Date(ctx.lastAlert.ts).getTime();
    const cooldownMs = ctx.cooldownHours * 3600 * 1000;
    if (deltaMs < cooldownMs) {
      const newRank = regimeRank(ctx.newRegime);
      const oldRank = regimeRank(safeRegime(ctx.lastAlert.regime));
      if (newRank <= oldRank) {
        return {
          allow: false,
          reason: `cooldown: same/lower regime ('${ctx.lastAlert.regime}' → '${ctx.newRegime}') within ${ctx.cooldownHours}h`,
        };
      }
    }
  }
  const sign = ctx.edgePct >= 0 ? "+" : "";
  return {
    allow: true,
    reason: `regime '${ctx.newRegime}', edge ${sign}${ctx.edgePct.toFixed(2)}%`,
  };
}

/**
 * edge = (current - baseline) / baseline * 100. Если baseline нет — 0.
 * Port src/alerts/gating.py:compute_edge_pct.
 */
export function computeEdgePct(currentRate: number, baselineMedian30d: number | null): number {
  if (baselineMedian30d === null || baselineMedian30d === 0) return 0;
  return ((currentRate - baselineMedian30d) / baselineMedian30d) * 100;
}

/**
 * Дневной edge — % изменения относительно предыдущей дневной свечи (вчера).
 * Считается из close-цен: (last.close - prev.close) / prev.close * 100.
 *
 * Возвращает null, если:
 * - меньше 2 daily candles (для новых assets без истории);
 * - prev.close == 0 (защита от деления на ноль).
 *
 * Используется в утреннем digest для строки «Edge за день», чтобы юзер видел
 * вчера→сегодня дельту рядом с месячным edge'ом.
 */
export function computeDailyEdgePct(daily: Candle[]): number | null {
  if (daily.length < 2) return null;
  const prev = daily[daily.length - 2].close;
  const last = daily[daily.length - 1].close;
  if (prev === 0 || Number.isNaN(prev) || Number.isNaN(last)) return null;
  return ((last - prev) / prev) * 100;
}

function safeRegime(value: string | null | undefined): Regime | null {
  if (value === "cooldown" || value === "watch" || value === "partial" || value === "strong") {
    return value;
  }
  return null;
}
